lib/commands/publish.js

```diff
@@ -0,0 +1,96 @@
+    { name: 'broker', type: String, description: 'URL of the Pact Broker to publish pacts to.', default: '' },
+    { name: 'consumer-version', type: String, description: 'A string containing a semver-style version, e.g. 1.0.0.', default: '' },
+    { name: 'broker-username', type: String, description: 'Username for Pact Broker basic authentication.', default: '' },
+    { name: 'broker-password', type: String, description: 'Password for Pact Broker basic authentication.', default: '' },
+    { name: 'tags', type: String, description: 'A comma-separated string to tag the Pacts being published.', default: '' },
+  ],
+
+  /*
+   * Generates a set of options that are passed to the pact client for
+   * publishing local pacts.
+   *
+   * @method getPublishOptions
+   * @private
+   * @param {String} projectRoot A path to the project to be able to default to
+   *                             create a default path to pacts
+   * @param {String} projectVersion Project version from `package.json` to be
+   *                                able to use it as a default consumer version during upload
+   * @param {Object} commandLineOptions A set of command line options defined
+   *                                    through `availableOptions`
+   */
+  getPublishOptions(projectRoot, projectVersion, commandLineOptions) {
+    let config = {
+      path: process.env.EMBER_CLI_PACT_PATH,
+      version: process.env.EMBER_CLI_PACT_CONSUMER_VERSION,
+      broker: process.env.EMBER_CLI_PACT_BROKER,
+      brokerUsername: process.env.EMBER_CLI_PACT_BROKER_USERNAME,
+      brokerPassword: process.env.EMBER_CLI_PACT_BROKER_PASSWORD,
+      tags: process.env.EMBER_CLI_PACT_TAGS
+    };
+    let mergedConfig = Object.assign({}, config, commandLineOptions);
+
+    let resultingConfig = {
+      pactFilesOrDirs: mergedConfig.path ? [mergedConfig.path] : [`${projectRoot}/pacts`],
+      consumerVersion: mergedConfig.version || projectVersion,
+      pactBroker: mergedConfig.broker || config.broker,
+      pactBrokerUsername: mergedConfig.brokerUsername,
+      pactBrokerPassword: mergedConfig.brokerPassword,
+      tags: mergedConfig.tags ? mergedConfig.tags.split(',') : []
+    };
+
+    if (!resultingConfig.pactBroker) {
+      throw new Error('URL of the Pact Broker is not specified! Please, use `.env` file or `--broker` parameter to set it.');
+    }
+
+    return resultingConfig;
+  },
+
+  /*
+   * A wrapper method around pact client to publish pacts to a pact broker.
+   *
+   * @method publishPacts
+   * @private
+   * @param {Object} options A set of publish options
+  */
+  publishPacts(options) {
+    let pact = require('@pact-foundation/pact-node');
+
+    return pact.publishPacts(options);
+  },
+
+  run(commandOptions/*, rawArgs*/) {
+    require('dotenv').config();
+
+    let projectRoot = this.project.root;
+    let projectVersion = this.project.pkg.version;
+    let publishingEnabled = process.env.EMBER_CLI_PACT_ENABLE_PUBLISH === 'true';
```

Comment:
I don't want our users to be in the position when they ran `publish` command and messed something up on the remote broker (either somehow incomplete pact, wrong tag, wrong version or possibly wrong consumer name). Pacts should generally be published from CI servers. From that standpoint, what is a compelling reason to publish locally unless you are extra super sure you want to do it?

But that's just my thinking, definitely open to discussion 😄

Reply:
Here's how I've been thinking about it: to be able to publish at all, you already need to pass in several pieces of configuration either via CLI flags or environment variables. I don't see having one extra environment variable that needs to be set as a barrier to entry that's likely to make someone think twice when they otherwise wouldn't have — either I'm running the command because I don't know what I'm doing (so I'm likely missing other config), or I'm running it on purpose and I've given exactly as much thought to setting `ENABLED` as I did to setting the username, password, etc. My 2 cents, anyway 🙂